src/Core/Package.vala

```diff
@@ -898,6 +898,67 @@ public class AppCenterCore.Package : Object {
         return null;
     }
 
+    public GenericArray<AppStream.Screenshot> get_screenshots () {
+        bool has_matching_environment = false;
+        bool has_matching_style = false;
+        var desktop_environment = Environment.get_variable ("XDG_SESSION_DESKTOP");
+        var prefer_dark_style = Gtk.Settings.get_default ().gtk_application_prefer_dark_theme;
+        var desktop_style = prefer_dark_style? AppStream.ColorSchemeKind.DARK.to_string () : AppStream.ColorSchemeKind.LIGHT.to_string ();
+
+        component.sort_screenshots (desktop_environment, desktop_style.to_string (), false);
+        var all_screenshots = component.get_screenshots_all ();
+
+        // This first pass is to gather if we have matching style and matching
+        // desktop environments, this is useful if we need to fall back to any
+        // screnshot if none of the conditions are fullfiled
+        foreach (unowned var screenshot in all_screenshots) {
+            var environment_id = screenshot.get_environment ();
+            if (environment_id != null) {
+                var environment_split = environment_id.split (":", 2);
+                var screenshot_environment = environment_split[0];
+                var screenshot_style = environment_split[1] ?? AppStream.ColorSchemeKind.LIGHT.to_string ();
+
+                if (screenshot_environment == desktop_environment) {
+                    has_matching_environment = true;
+                }
+
+                if (screenshot_style == desktop_style) {
+                    has_matching_style = true;
+                }
+            }
+        }
+
+        var screenshots = new GenericArray<AppStream.Screenshot> ();
+
+        foreach (unowned var screenshot in all_screenshots) {
+            var environment_id = screenshot.get_environment ();
+            if (environment_id == null) {
+                screenshots.add (screenshot);
+                break;
+            }
+
+            var environment_split = environment_id.split (":", 2);
+            var screenshot_environment = environment_split[0];
+            var screenshot_style = environment_split[1] ?? AppStream.ColorSchemeKind.LIGHT.to_string ();
+
+            var same_environment = screenshot_environment == desktop_environment;
+            var same_style = screenshot_style == desktop_style;
+
+            if (same_environment && same_style) {
+                screenshots.add (screenshot);
+            } else if (same_environment && !same_style && !has_matching_style) {
+                screenshots.add (screenshot);
+            } else if (!same_environment && same_style && !has_matching_environment) {
+                screenshots.add (screenshot);
+            } else if (!has_matching_environment && !has_matching_style) {
+                screenshots.add (screenshot);
+                break;
```

Comment:
This should also be `continue;` if you don't mean to change the meaning of the code with this PR. The previous code uses `GenericArray.@foreach ()` and returning in the function given as the argument is equivalent to call `continue;` in the `foreach (var elem in array)` statement.